File: api/src/test/java/net/blueshell/api/validation/SurveyValidatorTest.java
package net.blueshell.api.validation;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validator;
import net.blueshell.api.common.enums.QuestionType;
import net.blueshell.api.dto.survey.AnswerDTO;
import net.blueshell.api.dto.survey.QuestionDTO;
import net.blueshell.api.dto.survey.SurveyDTO;
import net.blueshell.api.factory.dto.survey.AnswerDTOFactory;
import net.blueshell.api.factory.dto.survey.QuestionDTOFactory;
import net.blueshell.api.factory.dto.survey.SurveyDTOFactory;
import net.blueshell.api.model.survey.Question;
import net.blueshell.api.repository.survey.QuestionRepository;
import net.blueshell.api.testutil.ModelTestUtils;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.bean.override.mockito.MockitoBean;

import java.util.List;
import java.util.Optional;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.when;

/**
 * Unit tests for Survey, Question, and Answer DTO validations.
 */
@SpringBootTest
class SurveyValidatorTest {

    @Autowired private Validator validator;
    @Autowired private SurveyDTOFactory surveyFactory;
    @Autowired private QuestionDTOFactory questionFactory;
    @Autowired private AnswerDTOFactory answerFactory;
    @MockitoBean private QuestionRepository questionRepository;

    private static Question mkQuestion(QuestionType type) {
        Question q = new Question();
        ModelTestUtils.setId(q, 42L);
        q.setType(type);
        if (type == QuestionType.RADIO || type == QuestionType.CHECKBOX) {
            q.setChoiceLabels(List.of("A", "B", "C"));
        }
        return q;
    }

    @Test
    void validSurveyDTO_passesValidation() {
        SurveyDTO dto = surveyFactory.createBasic();
        Set<ConstraintViolation<SurveyDTO>> violations = validator.validate(dto);
        assertTrue(violations.isEmpty(), "Valid SurveyDTO should pass validation");
    }

    @Test
    void surveyDTO_withEmptyQuestions_failsValidation() {
        SurveyDTO dto = surveyFactory.createWithCustomizations(s -> s.setQuestions(List.of()));
        Set<ConstraintViolation<SurveyDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("questions")));
    }

    @Test
    void validQuestionDTO_passesValidation() {
        QuestionDTO dto = questionFactory.createBasic();
        Set<ConstraintViolation<QuestionDTO>> violations = validator.validate(dto);
        assertTrue(violations.isEmpty(), "Valid QuestionDTO should pass validation");
    }

    @Test
    void questionDTO_withoutLabel_failsValidation() {
        QuestionDTO dto = questionFactory.createWithCustomizations(q -> q.setLabel(""));
        Set<ConstraintViolation<QuestionDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("label")));
    }

    @Test
    void questionDTO_withLongLabel_failsValidation() {
        String longLabel = "A".repeat(2056);
        QuestionDTO dto = questionFactory.createWithCustomizations(q -> q.setLabel(longLabel));
        Set<ConstraintViolation<QuestionDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("label")));
    }

    @Test
    void openQuestion_withChoiceLabels_failsValidation() {
        QuestionDTO dto = questionFactory.createWithCustomizations(q -> {
            q.setType(QuestionType.OPEN);
            q.setChoiceLabels(List.of("Choice 1", "Choice 2"));
        });
        Set<ConstraintViolation<QuestionDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
    }

    @Test
    void checkboxQuestion_withoutChoiceLabels_failsValidation() {
        QuestionDTO dto = questionFactory.createWithCustomizations(q -> {
            q.setType(QuestionType.CHECKBOX);
            q.setChoiceLabels(List.of());
        });
        Set<ConstraintViolation<QuestionDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
    }

    @Test
    void validAnswerDTO_passesValidation() {
        AnswerDTO dto = answerFactory.createBasic();
        when(questionRepository.findById(dto.getQuestionId()))
                .thenReturn(Optional.of(mkQuestion(QuestionType.OPEN)));

        Set<ConstraintViolation<AnswerDTO>> violations = validator.validate(dto);
        assertTrue(violations.isEmpty(), "Valid AnswerDTO should pass validation");
    }

    @Test
    void answerDTO_withoutQuestionId_failsValidation() {
        AnswerDTO dto = answerFactory.createWithCustomizations(a -> a.setQuestionId(null));
        Set<ConstraintViolation<AnswerDTO>> violations = validator.validate(dto);
        assertFalse(violations.isEmpty());
        assertTrue(violations.stream().anyMatch(v -> v.getPropertyPath().toString().equals("questionId")));
    }
}

File: api/src/test/kotlin/net/blueshell/api/survey/api/validation/SurveyValidatorTest.kt
package net.blueshell.api.survey.api.validation

import jakarta.validation.ConstraintViolation
import jakarta.validation.Validator
import net.blueshell.api.shared.enums.QuestionType
import net.blueshell.api.survey.api.dto.AnswerDTO
import net.blueshell.api.survey.api.dto.QuestionDTO
import net.blueshell.api.survey.api.dto.SurveyDTO
import net.blueshell.api.factory.dto.survey.AnswerDTOFactory
import net.blueshell.api.factory.dto.survey.QuestionDTOFactory
import net.blueshell.api.factory.dto.survey.SurveyDTOFactory
import net.blueshell.api.survey.domain.model.Question
import net.blueshell.api.survey.persistence.QuestionRepository
import net.blueshell.api.testutil.ModelTestUtils
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.mockito.kotlin.whenever
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.bean.override.mockito.MockitoBean
import java.util.Optional

/**
 * Unit tests for Survey, Question, and Answer DTO validations.
 */
@SpringBootTest
class SurveyValidatorTest @Autowired constructor(
    private val validator: Validator,
    private val surveyFactory: SurveyDTOFactory,
    private val questionFactory: QuestionDTOFactory,
    private val answerFactory: AnswerDTOFactory
) {

    @MockitoBean
    private lateinit var questionRepository: QuestionRepository

    private fun mkQuestion(type: QuestionType): Question {
        val question = Question()
        ModelTestUtils.setId(question, 42L)
        question.type = type
        if (type == QuestionType.RADIO || type == QuestionType.CHECKBOX) {
            question.choiceLabels = mutableListOf("A", "B", "C")
        }
        return question
    }

    @Test
    fun `valid survey dto passes validation`() {
        val dto = surveyFactory.createBasic()
        val violations: Set<ConstraintViolation<SurveyDTO>> = validator.validate(dto)
        assertTrue(violations.isEmpty(), "Valid SurveyDTO should pass validation")
    }

    @Test
    fun `survey dto with empty questions fails validation`() {
        val dto = surveyFactory.createWithCustomizations { it.questions = mutableListOf() }
        val violations: Set<ConstraintViolation<SurveyDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "questions" })
    }

    @Test
    fun `valid question dto passes validation`() {
        val dto = questionFactory.createBasic()
        val violations: Set<ConstraintViolation<QuestionDTO>> = validator.validate(dto)
        assertTrue(violations.isEmpty(), "Valid QuestionDTO should pass validation")
    }

    @Test
    fun `question dto without label fails validation`() {
        val dto = questionFactory.createWithCustomizations { it.label = "" }
        val violations: Set<ConstraintViolation<QuestionDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "label" })
    }

    @Test
    fun `question dto with long label fails validation`() {
        val longLabel = "A".repeat(2056)
        val dto = questionFactory.createWithCustomizations { it.label = longLabel }
        val violations: Set<ConstraintViolation<QuestionDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "label" })
    }

    @Test
    fun `open question with choice labels fails validation`() {
        val dto = questionFactory.createWithCustomizations {
            it.type = QuestionType.OPEN
            it.choiceLabels = mutableListOf("Choice 1", "Choice 2")
        }
        val violations: Set<ConstraintViolation<QuestionDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
    }

    @Test
    fun `checkbox question without choice labels fails validation`() {
        val dto = questionFactory.createWithCustomizations {
            it.type = QuestionType.CHECKBOX
            it.choiceLabels = mutableListOf()
        }
        val violations: Set<ConstraintViolation<QuestionDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
    }

    @Test
    fun `valid answer dto passes validation`() {
        val dto = answerFactory.createBasic()
        whenever(questionRepository.findById(dto.questionId)).thenReturn(Optional.of(mkQuestion(QuestionType.OPEN)))

        val violations: Set<ConstraintViolation<AnswerDTO>> = validator.validate(dto)
        assertTrue(violations.isEmpty(), "Valid AnswerDTO should pass validation")
    }

    @Test
    fun `answer dto without question id fails validation`() {
        val dto = answerFactory.createWithCustomizations { it.questionId = null }
        val violations: Set<ConstraintViolation<AnswerDTO>> = validator.validate(dto)
        assertFalse(violations.isEmpty())
        assertTrue(violations.any { it.propertyPath.toString() == "questionId" })
    }
}
